lexicon/providers/auto.py

```diff
@@ -0,0 +1,173 @@
+                         'Found nameservers domains are {1}'.format(domain, nameserver_domains))
+
+    if len(relevant_providers) > 1:
+        LOGGER.warn('Warning, multiple DNS providers have been found for given domain {0}, first one will be used: {1} '
+                    'This may indicate a misconfiguration in one or more provider.'
+                    .format(domain, relevant_providers))
+
+    return relevant_providers[0]
+
+def ProviderParser(subparser):
+    subparser.description = '''
+        Provider 'auto' enables the Lexicon provider auto-discovery feature.
+        Based on the nameservers declared for the given domain, Lexicon will try to find the DNS provider holding the DNS zone if it is supported.
+        Actual DNS zone read/write operations will be delegated to the provider found: every environment variable or command line specific to this provider can be passed to Lexicon and will be processed accordingly.
+        '''
+    subparser.add_argument("--mapping-override", metavar="[DOMAIN]:[PROVIDER], ...", help="comma separated list of elements in the form of [DOMAIN]:[PROVIDER] to authoritatively map a particular domain to a particular provider")
+
+    # Explore and load the arguments available for every provider into the 'auto' provider.
+    for provider_name, provider_module in AVAILABLE_PROVIDERS.items():
+        parser = argparse.ArgumentParser(add_help=False)
+        provider_module.ProviderParser(parser)
+
+        for action in parser._actions:
+            action.option_strings = [re.sub(r'^--(.*)$', r'--{0}-\1'.format(provider_name), option) for option in action.option_strings]
+            action.dest = 'auto_{0}_{1}'.format(provider_name, action.dest)
+            subparser._add_action(action)
+
+# Take care of the fact that this provider extends object, not BaseProvider !
+# Indeed we want to delegate every parameter/method call to the delegate provider
+# but __getattr__ is called only if the parameter/method cannot be found in the
+# current Provider hierarchy. If it is object, it will be the case for every relevant
+# call in the Lexicon library.
+class Provider(object):
+    """
+    Implementation of the provider 'auto'.
+    For the given domain, it will resolve the actual Provider class to use by inspecting the 
+    nameservers declared for the domain, using declared nameservers domain in each Lexicon DNS
+    provider.
+    Any call upon instantiation will be delegated to the resolved Provider.
+    Any command line parameter or environment variable passed to Lexicon will be transfered to
+    the resolved provider if it respect the naming convention: --[provider]-[parameter_name] for a
+    command line parameter, or LEXICON_[PROVIDER]_PARAMETER_NAME for a environment variable.
+    """
+    def __init__(self, options, engine_overrides=None):
+        self.domain = options.get('domain')
+        self.delegate = None
```

Comment:
another nit. we already use the term/parameter `delegate` elsewhere in relation to delegated domains. It might be better to use a different variable name here. 

possibly something like: `handler`, `primary`, `operator`, `proxy`? not a big deal though.

Reply:
I will tag `proxy_provider` to be explicit about the nature of the proxy. Next commit.